core/input/input_enums.h

```diff
@@ -47,6 +47,13 @@ enum class HatMask {
 	LEFT = 8,
 };
 
+enum class InputEventState {
+	NONE = 0,
```

Comment:
It would be nice to get rid of it. A way may be twisting a bit the naming so that `RELEASED` becomes `DEFAULT`. I'm open to better ideas.

Reply:
Seems like it could work having `{ DEFAULT = 0, PRESSED = 1, CANCELLED = 2 }`, we may need a better term instead of `DEFAULT` though.

Reply:
Looking at the documentation though, we have other enums with a `NONE` state (e.g: `KEY_NONE`, `MOUSE_BUTTON_NONE`, `MIDI_MESSAGE_NONE`) so it's still consistent keeping a `NONE` value for this enum.

Reply:
There may be a point in having null-like enum values in cases where the code may have to decide the appropriate value from some input data. However, in this case I quite don't see what usage `NONE` could have. I still strongly believe in the `DEFAULT, PRESSED, CANCELLED` triplet being the right approach here since no functionality at all needs to use `NONE` to flag anything and future pieces of engine or user code will have to pay the conginitive cost of reasoning about how to handle `NONE`.

Furthermore, if other areas of the engine are using a `NONE` value as a null-like for error situations, they should be refactored to tell the caller there's an error via a `bool`, `Err` or something like that.

That said, the consistency mindset has its own merits and avoids the deeper discussion about how to use enums in the engine. In short, do what you think is best.

Reply:
> I don't have a strong opinion about this, but `DEFAULT` being equivalent to released feels wrong. It feels like the "default state" of a key press or mouse click is pressed, not released.
> 
> If going for a triplet, perhaps `{ DEFAULT = 0, RELEASED = 1, CANCELLED = 2 }` (changing `PRESSED` to `RELEASED`) would be better?

@dsnopek The other approach (`{DEFAULT, PRESSED, CANCELLED}`) reflects our implementation and usage better; `InputEventMouseButton` and `InputEventScreenTouch` have for example their (previous) `pressed` field set to `false` by default, which would indicate `RELEASED` being the default state.

An alternate naming that may fit better could be `{INACTIVE = 0, ACTIVE = 1, CANCELLED = 2}`:
- For events with pressed state (e.g: `InputEventScreenTouch`, `InputEventMouseButton`), `ACTIVE` represents the `pressed` state, and `INACTIVE` represents the `released` state.
- For events without pressed state (e.g: `InputEventScreenDrag`, `InputEventMouseMotion`), they can default to always be in the `ACTIVE` state when created.

Reply:
I'd like to discourage the inactive/active/canceled idea, because, as I see it, it starts to get too abstract.

Reply:
> I'd like to discourage the inactive/active/canceled idea, because, as I see it, it starts to get too abstract.

That's my issue with using `DEFAULT` as well; it's a bit generic and not explicit enough as from your point of view it matches `RELEASED`, whereas from @dsnopek's point of view it matches `PRESSED`. This is likely to confuse users as well.

For now I'll stick to the `{NONE = 0, PRESSED = 1, RELEASED = 2, CANCELED = 3}` approach, as it covers all scenarios and is consistent with other enums in the codebase, but I'm opened to revisit if we can all agree on an alternative.

Reply:
Yeah. Makes sense. Let's stop the bikeshedding!